packages/flutter/lib/src/material/tabs.dart

```diff
@@ -760,543 +352,478 @@ class TabBar<T> extends Scrollable implements AppBarBottomWidget {
-    children: children
-  ) {
-    assert(children != null);
-    assert(children.length > 1);
-  }
+    children: children,
+    initialScrollOffset: initialScrollOffset,
+  );
 
   @override
-  _TabBarViewState<T> createState() => new _TabBarViewState<T>();
+  _PageableTabBarViewState createState() => new _PageableTabBarViewState();
 }
 
-class _TabBarViewState<T> extends PageableListState<TabBarView<T>> implements TabBarSelectionAnimationListener {
-
-  TabBarSelectionState<T> _selection;
-  List<Widget> _items;
-
-  int get _tabCount => config.children.length;
-
+class _PageableTabBarViewState extends PageableListState<_PageableTabBarView> {
   BoundedBehavior _boundedBehavior;
 
   @override
   ExtentScrollBehavior get scrollBehavior {
-    _boundedBehavior ??= new BoundedBehavior(platform: platform);
+    _boundedBehavior ??= new BoundedBehavior(
+      platform: platform,
+      containerExtent: 1.0,
+      contentExtent: config.children.length.toDouble(),
+    );
     return _boundedBehavior;
   }
 
   @override
   TargetPlatform get platform => Theme.of(context).platform;
 
-  void _initSelection(TabBarSelectionState<T> newSelection) {
-    if (_selection == newSelection)
-      return;
-    _selection?.removeAnimationListener(this);
-    _selection = newSelection;
-    _selection?.addAnimationListener(this);
-    if (_selection != null)
-      _updateItemsAndScrollBehavior();
+  @override
+  Future<Null> fling(double scrollVelocity) {
+    final double newScrollOffset = snapScrollOffset(scrollOffset + scrollVelocity.sign)
+      .clamp(snapScrollOffset(scrollOffset - 0.5), snapScrollOffset(scrollOffset + 0.5))
+      .clamp(0.0, (config.children.length - 1).toDouble());
+    return scrollTo(newScrollOffset, duration: config.duration, curve: config.curve);
   }
 
   @override
-  void didUpdateConfig(TabBarView<T> oldConfig) {
-    super.didUpdateConfig(oldConfig);
-    if (_selection != null && config.children != oldConfig.children)
-      _updateItemsForSelectedIndex(_selection.index);
+  Widget buildContent(BuildContext context) {
+    return new PageViewport(
+      mainAxis: config.scrollDirection,
+      startOffset: scrollOffset,
+      children: config.children,
+    );
   }
+}
 
-  @override
-  void dispose() {
-    _selection?.removeAnimationListener(this);
-    super.dispose();
+/// A pageable list that displays the widget which corresponds to the currently
+/// selected tab. Typically used in conjuction with a [TabBar].
+///
+/// If a [TabController] is not provided, then there must be a [DefaultTabController]
+/// ancestor.
+class TabBarView extends StatefulWidget {
+  /// Creates a pageable list with one child per tab.
+  ///
+  /// The length of [children] must be the same as the [controller]'s length.
+  TabBarView({
+    Key key,
+    this.children,
+    this.controller,
+  }) : super(key: key) {
+    assert(children != null && children.length > 1);
   }
 
+  /// This widget's selection and animation state.
```

Comment:
I'd repeat the doc about this defaulting to the nearest enclosing DefaultTabController if null.